lib/kinesis-client.js

```diff
@@ -0,0 +1,224 @@
+      const error = new Error(message);
+      error.code = code;
+      throw error;
+    });
+  }
+
+  listShards(...args) {
+    const { client, retryOpts } = internal(this);
+    return retry(
+      bail =>
+        client
+          .listShards(...args)
+          .promise()
+          .catch(err => {
+            if (shouldBailRetry(err)) bail(err);
+            else throw err;
+          }),
+      retryOpts
+    ).catch(err => {
+      const { code, message } = err;
+      const error = new Error(message);
+      error.code = code;
+      throw error;
+    });
+  }
+
+  listTagsForStream(...args) {
+    const { client, retryOpts } = internal(this);
+    return retry(
+      bail =>
+        client
+          .listTagsForStream(...args)
+          .promise()
+          .catch(err => {
+            if (shouldBailRetry(err)) bail(err);
+            else throw err;
+          }),
+      retryOpts
+    ).catch(err => {
+      const { code, message } = err;
+      const error = new Error(message);
+      error.code = code;
+      throw error;
+    });
+  }
+
+  startStreamEncryption(...args) {
+    return internal(this)
+      .client.startStreamEncryption(...args)
+      .promise()
+      .catch(err => {
+        const { code, message } = err;
+        if (code !== 'UnknownOperationException' && code !== 'ResourceInUseException') {
+          const error = new Error(message);
+          error.code = code;
+          throw error;
+        }
+      });
+  }
+
+  waitFor(...args) {
+    const { client, retryOpts } = internal(this);
+    return retry(
+      bail =>
+        client
+          .waitFor(...args)
+          .promise()
+          .catch(err => {
+            if (shouldBailRetry(err)) bail(err);
+            else throw err;
+          }),
+      retryOpts
+    ).catch(err => {
+      const { code, message } = err;
+      const error = new Error(message);
+      error.code = code;
+      throw error;
+    });
+  }
+
+  getShardIterator(...args) {
+    const { client, retryOpts } = internal(this);
+    return retry(
+      bail =>
+        client
+          .getShardIterator(...args)
+          .promise()
+          .catch(err => {
+            if (shouldBailRetry(err)) bail(err);
+            else throw err;
+          }),
+      retryOpts
+    ).catch(err => {
+      const { code, message } = err;
+      const error = new Error(message);
+      error.code = code;
+      throw error;
+    });
+  }
+
+  getRecords(...args) {
+    const { client, retryOpts } = internal(this);
+    return retry(
+      bail =>
+        client
+          .getRecords(...args)
+          .promise()
+          .catch(err => {
+            if (shouldBailRetry(err)) bail(err);
+            else throw err;
+          }),
+      retryOpts
+    ).catch(err => {
+      const { code, message } = err;
+      const error = new Error(message);
+      error.code = code;
+      throw error;
+    });
+  }
+
+  putRecord(...args) {
+    return internal(this)
```

Comment:
Do we want to add any retry logic to these putRecord(s) ? Just something to think about.